src/wodan/keyedmap.mli

```diff
@@ -0,0 +1,115 @@
+      is greater than [k], or raises [Not_found] if empty *)
+
+  val find_last : 'a t -> key -> key * 'a
+  (** [find_first_opt m k] finds the binding with the greatest key that
+      is smaller than [k], or raises [Not_found] if empty *)
+
+  val find_first_opt : 'a t -> key -> (key * 'a) option
+  (** Same as [find_first] but wrapped in an option *)
+
+  val find_last_opt : 'a t -> key -> (key * 'a) option
+  (** Same as [find_last] but wrapped in an option *)
+
+  val mem : 'a t -> key -> bool
+  (** [mem m x] checks whether [x] is bind in [m]  *)
+
+  val exists : (key -> 'a -> bool) -> 'a t -> bool
+  (** [exists f m] checks whether there exists a binding in [m]
+      satisfying the predicate [f] *)
+
+  val add : 'a t -> key -> 'a -> unit
+  (** [add m k v] adds a binding from [k] to [v] in [m], replacing
+      the previous binding if any*)
+
+  val remove : 'a t -> key -> unit
+  (** [remove m k] removes the binding of [k] in [m] *)
+
+  val replace_existing : 'a t -> key -> 'a -> unit
+  (** [replace_existing m k v] replaces the binding of [k] in
+      [m] with the value [v] if any, otherwise raises [Not_found] *)
+
+  val xadd : 'a t -> key -> 'a -> unit
+  (** [xadd m k v] adds a binding from [k] to [v], or raises
+      [Already_exists] if there is already one *)
+
+  val update : 'a t -> key -> ('a option -> 'a option) -> unit
+  (** [update m x f] returns a map containing the same bindings as [m], except
+      for the binding of [x]. Depending on the value of y where y is
+      [f (find_opt x m)], the binding of [x] is added, removed or updated.
+      If y is [None], the binding is removed if it exists; otherwise, if y is
+      [Some z] then [x] is associated to [z] in the resulting map. If [x] was
+      already bound in [m] to a value that is physically equal to [z], [m] is
+      left unchanged *)
+
+  val keys : 'a t -> key list
+  (** Returns the list of the keys of the map, in ascending order *)
+
+  val iter : (key -> 'a -> unit) -> 'a t -> unit
+  (** [iter f m] applies [f] to all bindings in [m] *)
+
+  val fold : (key -> 'a -> 'b -> 'b) -> 'a t -> 'b -> 'b
+  (** [fold f m a] computes [(f kN dN ... (f k1 d1 a)...)], where [k1 ... kN]
+      are the keys of all bindings in [m] (in increasing order), and
+      [d1 ... dN] are the associated data. *)
+
+  val iter_range : (key -> 'a -> unit) -> 'a t -> key -> key -> unit
+  (** [iter f m start stop] applies [f] to all bindings [(x, d)] in [m] such
+      that start <= x < stop *)
+
+  val iter_inclusive_range : (key -> 'a -> unit) -> 'a t -> key -> key -> unit
+  (** [iter f m start stop] applies [f] to all bindings [(x, d)] in [m] such
+      that start <= x <= stop *)
+
+  val split_off_after : 'a t -> key -> 'a t
+  (** [split_off_after m k] removes all the bindings [(x, v)] in [m] such
+      that [x > k] and returns them in a new map *)
+
+  val carve_inclusive_range : 'a t -> key -> key -> (key * 'a) list
```

Comment:
I'm not sure it's better to return a list here.
The underlying partitioning function works with maps, I think that's preferable.
If we ever go with a more optimised KeyedMap implementation, it's also simpler to implement everything with the same types (tested in the Rust case).

Reply:
Okay, this came to me when I realized that we only `iter` through the results of this function in Wodan, so a `list` felt right, but I don't really mind changing it back if you think that's better.